# TP 2-3 : Branch-and-bound applied to a knapsack problem

### Initialisation (à faire une seule fois)

In [1]:
import Pkg; 
Pkg.add("GraphRecipes"); Pkg.add("Plots"); 
using GraphRecipes, Plots #only used to visualize the search tree at the end of the branch-and-bound

    Updating registry at `C:\Users\bonet\.julia\registries\General`


    Updating git-repo `https://github.com/JuliaRegistries/General.git`


   Resolving package versions...


  No Changes to `C:\Users\bonet\.julia\environments\v1.8\Project.toml`
  No Changes to `C:\Users\bonet\.julia\environments\v1.8\Manifest.toml`

   Resolving package versions...


  No Changes to `C:\Users\bonet\.julia\environments\v1.8\Project.toml`
  No Changes to `C:\Users\bonet\.julia\environments\v1.8\Manifest.toml`

### Récupération des données

In [2]:
function readKnaptxtInstance(filename)
    price=[]
    weight=[]
    KnapCap=[]
    open(filename) do f
        for i in 1:3
            tok = split(readline(f))
            if(tok[1] == "ListPrices=")
                for i in 2:(length(tok)-1)
                    push!(price,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "ListWeights=")
                for i in 2:(length(tok)-1)
                    push!(weight,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "Capacity=")
                push!(KnapCap, parse(Int64, tok[2]))
            else
                println("Unknown read :", tok)
            end 
        end
    end
    capacity=KnapCap[1]
    return price, weight, capacity
end

readKnaptxtInstance (generic function with 1 method)

In [3]:
function creerTableau(prix, poids, capacite)
    n = length(prix)
    tableau = zeros(Int64, n, capacite + 1)
    # Initialisation de la première ligne
    tableau[1,1:poids[1]] .= 0
    tableau[1,(poids[1]+1):end] .= prix[1]
    for i in 2:n
        for j in 1:(capacite + 1)
            # Si le poids de l'objet est supérieur à la capacité, on prend la valeur de la ligne précédente
            if poids[i] > j-1 
                tableau[i,j] = tableau[i-1,j]
            # Sinon, on prend le maximum entre la valeur de la ligne précédente et la valeur de l'objet + la valeur de la ligne précédente à la capacité - le poids de l'objet
            else
                tableau[i,j] = max(tableau[i-1,j],prix[i] + tableau[i-1,j-poids[i]])
            end
        end
    end
    return tableau
end


creerTableau (generic function with 1 method)

In [4]:

function SolveKnapInstance(filename)

    price, weight, capacity = readKnaptxtInstance(filename)

    tableau = creerTableau(price, weight, capacity)

    solution = tableau[end,end]

    return solution

end


SolveKnapInstance (generic function with 1 method)

In [6]:
println(SolveKnapInstance("instancesETU/KNAPnewformat/test.opb.txt"))

for file in cd(readdir, "instanceTest/")
    println(file, " : ", SolveKnapInstance("instanceTest/"*file))
end

65


knapPI_16_20_1000_1_-2291.opb.txt : 2291
knapPI_16_20_1000_3_-2189.opb.txt : 3267
knapPI_16_20_1000_5_-3230.opb.txt : 

4437


knapPI_4_50_10000_5_-11993.opb.txt : 12005
knapPI_4_50_1000_1_-994.opb.txt : 994


knapPI_4_50_1000_5_-1278.opb.txt : 1280


knapPI_5_50_10000_2_-18843.opb.txt : 24908


knapPI_5_50_10000_4_-18105.opb.txt : 32113
knapPI_5_50_1000_1_-2096.opb.txt : 2096


knapPI_5_50_1000_4_-1872.opb.txt : 2977


 ## Réponses aux questions 
 
 1. Implémentation
   
 2. Le résultat est en adéquation avec la solution optimale car l'algorithme décompose le problème en étapes et chaque étape corresponds à un sous problème résolu optimalement grâce à une relation de récurrence entre les infos de l'étape courante et celles obtenues lors des étapes précédentes. Ainsi le résultat de chaque étape est optimal et le résultat final est optimal.
   
 3. Notre algorithme est décomposé en deux fonctions.
   
La première fonction prend en argument trois paramètres :

  -prix est un tableau des valeurs des objets
  -poids est un tableau des poids des objets
  -capacite est la capacité du sac à dos

Elle crée un tableau à deux dimensions, avec des lignes correspondant aux objets et des colonnes correspondant aux capacités possibles du sac à dos. La tableau est initialisé avec des zéros, puis les valeurs de la première ligne sont remplies en utilisant un cas spécial pour le premier objet. (Si le poids de l'objet est supérieur à la capacité, la valeur est zéro, sinon elle est égale à la valeur de l'objet).

La fonction parcourt ensuite les objets restants et, pour chaque objet, parcourt toutes les capacités possibles du sac à dos. Pour chaque capacité, elle calcule la valeur maximale qui peut être obtenue en ne prenant pas l'objet actuel ou en prenant l'objet actuel et en soustraisant son poids de la capacité. Le maximum de ces deux valeurs est alors enregistré dans le tableau.

La fonction retourne le tableau ainsi créé.

La deuxième fonction recupère les données du problème puis utilise la première fonction pour créer le tableau et enfin renvoie la dernière case tu tableau qui correspond à la solution.

4. On test pour différentes instances de tailles différentes les deux algorithme, celui du TP2 basé sur le branch and bound et celui du TP3 basé la programmation dynamique. Pour l'instance exemple du TP2, les deux algorithme trouvent bien 65, ce qui est logique puisque l'instance possède un jeu de données minime. 

Voici les résultats pour différentes instances : 

knapPI_16_20_1000_1_-2291.opb.txt : 2291 $\newline$
knapPI_16_20_1000_3_-2189.opb.txt : 3267 $\newline$
knapPI_16_20_1000_5_-3230.opb.txt :4437 $\newline$
knapPI_4_50_10000_5_-11993.opb.txt : 12005  $\newline$
knapPI_4_50_1000_1_-994.opb.txt : 994 $\newline$
knapPI_4_50_1000_5_-1278.opb.txt : 1280 $\newline$
knapPI_5_50_10000_2_-18843.opb.txt : 24908 $\newline$
knapPI_5_50_10000_4_-18105.opb.txt : 32113 $\newline$
knapPI_5_50_1000_1_-2096.opb.txt : 2096 $\newline$
knapPI_5_50_1000_4_-1872.opb.txt : 2977

De son côté, l'algorithme que nous avons implémenté pour le TP2 donne des résultats incohérents pour la plupart de nos tests, ce qui est sans doute dû à une erreur dans l'implémentation de l'algorithme.

De manière générale, il semblerait que la méthode par programmation dynamique est plus efficace que la méthode de recherche approfondie de branch and bound pour des problèmes avec des jeux de données importants. En revanche, pour des jeux de données plus petits, la méthode de recherche approfondie est plus efficace. Ces observations sont en accord avec les calculs de complexité théorique des deux algorithmes : O(n*m) pour la programmation dynamique et O(2^n) pour la recherche approfondie.
